### ISSUE TITLE
vg map fails with sdsl error

### ISSUE THREAD

**Issue by psritharan** [vgteam/vg#1269]
I am currently using `vg map` to map reads from a yeast strain to my vg graph using the following command:
`vg map -if NCYC1006_trimmed_sorted_interleaved.fastq -x wholegenome.xg -g wg.gcsa > ncyc1006_aln.gam`

However, this causes vg map to immediately abort with the following error:
```
vg/include/sdsl/select_support_mcl.hpp:349: sdsl::select_support::size_type sdsl::select_support_mcl<t_bit_pattern, t_pattern_len>::select(sdsl::select_support::size_type) const [with unsigned char t_b = 1u; unsigned char t_pat_len = 1u; sdsl::select_support::size_type = long unsigned int]: Assertion `i > 0 and i <= m_arg_cnt' failed.
Aborted
```
I am currently using the latest build of vg on my server.  Any ideas on what could be causing this error? 
Thanks.

**Comment by adamnovak** [comment]
I suspect it's having trouble reading something out of the xg index. How big are the files you are having trouble with. Is there any chance of you being able to share them?

**Comment by psritharan** [comment]
Thanks for the quick reply. I followed the wiki example on working with whole genome variation graphs to construct the graphs. I have attached the reference graph (split up per chromosomes), xg index and gcsa index files for download [here](https://files.fm/u/d9mv444y). The interleaved fastq file containing the reads for mapping is too large to upload. 

The xg index file was created by generating the xg index for each graph separately then joint together as trying to generate the xg index across the graphs using : 
```
vg index -x wholegenome.xg $(for chr in *.vg; do echo $chr; done)
```
produces the following error:
```
[xg] warning: path S288c_ChrI is not in sorted order by rank
[xg] error: path S288c_ChrI contains duplicate node ranks
```

**Comment by adamnovak** [comment]
OK, it looks like you've merged multiple XG indexes by concatenation, which won't work and produces an invalid file. Concatenation works fine for .vg graphs, but not for .xg indexes. There is currently no way to merge XG indexes; you have to construct the index all at once from all the vg graphs for all the parts.

Your whole-genome index operation is failing, I think, because you have multiple paths named "S288c_ChrI" in different graphs, and you are trying to put them all in one index. Path names need to be unique across all the graphs. You can use `vg paths --list whatever.vg` to get the names of all the paths in a graph; try that on all the vg files and you will probably see two or more with a path called "S288c_ChrI".

Actually, it looks like your graphs all have all the chromosomes in them:

```
$ vg paths --list haploidmerged.S288c_ChrXIV.vg 
S288c_ChrI
S288c_ChrII
S288c_ChrIII
S288c_ChrIV
S288c_ChrIX
S288c_ChrV
S288c_ChrVI
S288c_ChrVII
S288c_ChrVIII
S288c_ChrX
S288c_ChrXI
S288c_ChrXII
S288c_ChrXIII
S288c_ChrXIV
S288c_ChrXV
S288c_ChrXVI
S288c_Chrmt
```

You can break up the graph into connected components with `vg explode`:

```
$ vg explode haploidmerged.S288c_ChrXIV.vg exploded
exploded/component0.vg	S288c_ChrXIV
exploded/component1.vg	S288c_ChrI
exploded/component2.vg	S288c_ChrII
exploded/component3.vg	S288c_ChrIII
exploded/component4.vg	S288c_ChrIV
exploded/component5.vg	S288c_ChrIX
exploded/component6.vg	S288c_ChrV
exploded/component7.vg	S288c_ChrVI
exploded/component8.vg	S288c_ChrVII
exploded/component9.vg	S288c_ChrVIII
exploded/component10.vg	S288c_ChrX
exploded/component11.vg	S288c_ChrXI
exploded/component12.vg	S288c_ChrXII
exploded/component13.vg	S288c_ChrXIII
exploded/component14.vg	S288c_ChrXV
exploded/component15.vg	S288c_ChrXVI
exploded/component16.vg	S288c_Chrmt
```

Ad then you can look at the node and edge counts for each with `vg stats`:

```
$ vg stats -z exploded/component16.vg 
nodes	86
edges	85
$ vg stats -z exploded/component0.vg 
nodes	25041
edges	35637
```

A graph with 1 more node than edge is probably just a long stick with no variants.

I think you've built graphs that include the variants for one chromosome but the reference versions of all the other chromosomes. You need to restrict your individual `vg construct` calls to just the chromosome they are supposed to be building by using the `-R/--region` option. Otherwise they'll all operate on everything they can find in the FASTA file, whether they have a VCF for it or not.

**Comment by psritharan** [comment]
Thank you!! I reconstructed the graphs with the -R parameter which worked and I was able to generate the xg index.

**Comment by adamnovak** [comment]
Excellent! I think that fixes the issue. Please reopen it if it still doesn't work.

**Comment by ChriKub** [comment]
Hi,

I'm running into the same problem when mapping reads to my graph.

`Loading xg index sevenref_Chr1.patched.xg...
Loading GCSA2 index sevenref_Chr1.patched.gcsa...
Loading LCP index sevenref_Chr1.patched.gcsa...
[vg map] : min_mem_length = 20, mem_reseed_length = 30, min_cluster_length = 0
align_paired_multi_simul with HWI-M00333R1:37:000000000-ANWUW:1:1101:1756:15719 and HWI-M00333R1:37:000000000-ANWUW:1:1101:1756:15719
fragment model 5000, 0, 0, 0, 0, 1, 0,
maybe_mq1 HWI-M00333R1:37:000000000-ANWUW:1:1101:1756:15719 1.43041 512 22 21 512 0.565737 0 132.17 512
maybe_mq2 HWI-M00333R1:37:000000000-ANWUW:1:1101:1756:15719 3.31252 512 21 19 512 0.163347 0 132.17 512
mems for read 1 [["CACTGTACATTGAGATGCAT",["2283632:645"]],["TGTACATTGAGATGCATTTT",["1512495:6"]],["TTTTTAAAAATATTTTGGTA",["1202780:-1"]],["TATTTTGGTATGGGGTATTT",["275351:-185"]],["AGCTATTGTTAAAAAAAAAAAC",["144466:-28"]],["TTTTTCAAAACTTTCCCTAT",["3091125:0"]],["TGATTATTATGACTAGTTCA",["1151592:-21","2911854:-0","2920416:-12","2924333:-2","2929188:-5","2932406:-8"]]]
mems for read 2 [["TATTTATATTTTCGAAGAACT",["279179:42"]],["AAATTTATTATCTATTTTTT",["1744493:1","1744517:1"]]]
vg: /ebio/abt6_projects9/abt6_software/bin/vg/include/sdsl/select_support_mcl.hpp:349: sdsl::select_support::size_type sdsl::select_support_mcl<t_bit_pattern, t_pattern_len>::select(sdsl::select_support::size_type) const [with unsigned char t_b = 1u; unsigned char t_pat_len = 1u; sdsl::select_support::size_type = long unsigned int]: Assertion `i > 0 and i <= m_arg_cnt' failed.`

The graph I build contains a multiple alignment of a single chromosome. `vg paths -L` returns every path exactly once and `vg explode` results in a copy of the previous file, so no separate subgraph is present.
The problem might be, that the graph is very large:

> nodes   3091124
edges   3566561
length  75538937
 
I'm currently trying to find the smallest possible subgraph that still returns the rerror, as the current files are way o big for sharing.

**Comment by ekg** [comment]
Are you certain that the gcsa2 and xg indexes are in sync?

**Comment by ChriKub** [comment]
Yes. They were created in the same `vg index` call. This should make them sync.

**Comment by ChriKub** [comment]
I created a subset that recreates the error message. The graph and fastq files can be downloaded [here](https://owncloud.tuebingen.mpg.de/index.php/s/lV2jWQC6w6dZTgm)
`vg index -x sdsl_error.xg -g sdsl_error.gcsa -k 11 -t 40 sdsl_error.vg`

`vg map -D -x sdsl_error.xg -g sdsl_error.gcsa -f reads1.fastq -f read2.fastq -t 40 sdsl_error.vg > sdsl_error.gam 2> sdsl_error.gout`

**Comment by ekg** [comment]
@ChriKub I cannot reproduce this on the current master. Is this still a problem?

**Comment by ChriKub** [comment]
@ekg Me neither. I'm happily mapping reads. I think this is solved. Thanks.

**Comment by faithokamoto** [comment]
Seems this was solved at the time.
